crates/net/common/src/bandwidth_monitor.rs

```diff
@@ -0,0 +1,176 @@
+//! Support for monitoring bandwidth. Takes heavy inspiration from https://github.com/libp2p/rust-libp2p/blob/master/src/bandwidth.rs
+
+// Copyright 2019 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::{
+    convert::TryFrom as _,
+    io,
+    pin::Pin,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+    task::{ready, Context, Poll},
+};
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+
+
+/// Monitors bandwidth usage of TCP streams
+pub struct BandwidthMonitor {
+    /// Measures the number of inbound packets
+    inbound: AtomicU64,
+    /// Measures the number of outbound packets
+    outbound: AtomicU64,
```

Comment:
let's move these to a new type `BandwidthMeterInner`

then this will be `BandwidthMeter {Arc<BandwidthMeterInner>}` 

so we can clone it and it's more convenient to handle

Reply:
Is it necessary to define `BandwidthMeter {Arc<BandwidthMeterInner>}` separately from the `MonitoredStream` struct I already have?

Reply:
for reference:
```rust
/// Wraps around a single stream that implements [`AsyncRead`] + [`AsyncWrite`] and monitors the bandwidth through it
#[pin_project::pin_project]
pub(crate) struct MonitoredStream<S> {
    /// The stream this instruments
    #[pin]
    inner: S,
    /// The [`BandwidthMonitor`] struct this uses to monitor bandwidth
    monitor: Arc<BandwidthMeterInner>,
}
```

Reply:
hmm actually yeah it'd be nice to pass around the monitor w/o the stream, i'll make a type alias for it

Reply:
I'd be nice to have this:

```rust
/// Monitors bandwidth usage of TCP streams
pub struct BandwidthMeterInner {
    /// Measures the number of inbound packets
    inbound: AtomicU64,
    /// Measures the number of outbound packets
    outbound: AtomicU64,
    
}

#[derive(Clone)]
pub struct BandwidthMeter {
    inner: Arc<BandwidthMeterInner>
}
```

This is basically a wrapper type around `Arc` but it's easier to manage than having a field : Arc<...>

and moving the getters to `BandwidthMeter`

Reply:
Ah I see, yup will make the change!
I had originally went for a `type BandwidthMeter = Arc<BandwidthMeterInner>`, I guess the `struct` approach is better for the reason of defining getters on it, right?
Just trying to get better at Rust heh

Reply:
there are two aspects:

* Arc<T> makes it more obvious that this is an Arc value but it's harder to use as a type
* `Strut{Arc<Inner>}` is easier to use but hides that this is just an arc

imo the latter is just more pleasant to work with